src/reactive/keystone_saml_mellon_handlers.py

```diff
@@ -67,69 +39,46 @@ def keystone_departed():
         charm_instance.remove_config()
 
 
-@reactive.when('endpoint.keystone-fid-service-provider.joined')
-@reactive.when_not('config.complete')
-def config_changed():
+@reactive.when('keystone-fid-service-provider.connected')
+def publish_sp_fid():
+    # don't always have a relation context - obtain from the flag
+    fid_sp = endpoint_from_flag(
+        keystone_saml_mellon.KEYSTONE_FID_ENDPOINT)
     with charm.provide_charm_instance() as charm_instance:
-        if charm_instance.configuration_complete():
-            flags.set_flag('config.complete')
-
-
-@reactive.when('endpoint.keystone-fid-service-provider.joined')
-@reactive.when_not('keystone-data.complete')
-def keystone_data_changed(fid_sp):
-    primary_data = fid_sp.all_joined_units[0].received
-    if primary_data:
-        hostname = primary_data.get('hostname')
-        port = primary_data.get('port')
-        tls_enabled = primary_data.get('tls-enabled')
-        # a basic check on the fact that keystone provided us with
-        # hostname and port information
-        if hostname and port:
-            # save hostname and port data in local storage for future
-            # use - in case config is incomplete but a relation is
-            # we need to store this across charm hook invocations
-            unitdb = unitdata.kv()
-            unitdb.set('hostname', hostname)
-            unitdb.set('port', port)
-            unitdb.set('tls-enabled', tls_enabled)
-            flags.set_flag('keystone-data.complete')
+        fid_sp.publish(charm_instance.options.protocol_name,
+                       charm_instance.options.remote_id_attribute)
 
 
-@reactive.when('endpoint.keystone-fid-service-provider.joined')
-@reactive.when('config.complete')
-@reactive.when('keystone-data.complete')
-@reactive.when_not('config.rendered')
+@reactive.when('keystone-fid-service-provider.available')
 def render_config():
     # don't always have a relation context - obtain from the flag
     fid_sp = endpoint_from_flag(
-        'endpoint.keystone-fid-service-provider.joined')
+        keystone_saml_mellon.KEYSTONE_FID_ENDPOINT)
     with charm.provide_charm_instance() as charm_instance:
-        charm_instance.render_config()
-        flags.set_flag('config.rendered')
-        # Trigger keystone restart. The relation is container-scoped
-        # so a per-unit db of a remote unit will only contain a nonce
-        # of a single subordinate
-        restart_nonce = str(uuid.uuid4())
-        fid_sp.publish(restart_nonce,
-                       charm_instance.options.protocol_name,
-                       charm_instance.options.remote_id_attribute)
-
-
-@reactive.when('endpoint.websso-fid-service-provider.joined')
-@reactive.when('config.complete')
-@reactive.when('keystone-data.complete')
-@reactive.when('config.rendered')
+        if charm_instance.configuration_complete():
+            print("COMPLETE")
```

Comment:
Should we have `print` functions in charms?  Is this left-over debug?

Reply:
No, that was accidentally added to git. Fixed.